### PR TITLE
feat: implement rocks transaction

### PR DESCRIPTION
**Pull request by bsbds**
Depends-On: #674 

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

  Implement txn API for rocksdb engine.

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

**Comment by Phoenix500526** [comment]
@Mergifyio rebase

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 1]
```diff
@@ -1,18 +1,395 @@
...
+/// Transaction type for `RocksDB`
+#[derive(Debug)]
+pub struct RocksTransaction {
+    /// Inner state
+    inner: Mutex<Option<Inner>>,
+}
+
+/// Inner state of the transaction
+///
+/// WARN: `db` should never be dropped before `txn`
```
Comment (Phoenix500526) [inline crates/engine/src/rocksdb_engine/transaction.rs:30]: Is it a gentleman agreement? Do we have any method to ensure that this warning will never happen?

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 2]
```diff
@@ -1,18 +1,395 @@
...
+            txn_size: 0,
+        };
+        Self {
+            inner: Mutex::new(Some(inner)),
+        }
+    }
+}
+
+#[allow(clippy::unwrap_used)]
+#[allow(clippy::unwrap_in_result)]
```
Comment (Phoenix500526) [inline crates/engine/src/rocksdb_engine/transaction.rs:93]: Please merge these two `allow` together. BTW, why do you allow them?

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 3]
```diff
@@ -1,18 +1,395 @@
...
+        &self,
+        table: &str,
+        keys: &[impl AsRef<[u8]>],
+    ) -> Result<Vec<Option<Vec<u8>>>, EngineError> {
+        self.inner.lock().as_mut().unwrap().get_multi(table, keys)
+    }
+}
+
+#[allow(clippy::unwrap_used)]
+#[allow(clippy::unwrap_in_result)]
```
Comment (Phoenix500526) [inline crates/engine/src/rocksdb_engine/transaction.rs:117]: Same as above

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 4]
```diff
@@ -1,18 +1,395 @@
...
+
+    fn rollback(&self) -> Result<(), EngineError> {
+        self.inner.lock().as_mut().unwrap().rollback()
+    }
+}
+
+impl Inner {
+    /// Replace txn with a new transaction
+    #[allow(unsafe_code)]
+    #[allow(clippy::unwrap_used)]
```
Comment (Phoenix500526) [inline crates/engine/src/rocksdb_engine/transaction.rs:131]: Same as above.

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 5]
```diff
@@ -1,18 +1,395 @@
...
+    fn enable_transaction(&mut self) -> Result<(), EngineError> {
+        if self.txn.is_some() {
+            return Ok(());
+        }
+
+        let txn = self.db.transaction();
+        let txn_static =
```
Comment (Phoenix500526) [inline crates/engine/src/rocksdb_engine/transaction.rs:138]: Is it a must to use unsafe code?

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 6]
```diff
@@ -1,18 +1,395 @@
...
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[allow(clippy::unwrap_used)]
+#[allow(clippy::unwrap_in_result)]
```
Comment (Phoenix500526) [inline crates/engine/src/rocksdb_engine/transaction.rs:243]: Same as above

File: crates/engine/src/rocksdb_engine/transaction.rs (hunk 1 above)
Comment (bsbds) [reply crates/engine/src/rocksdb_engine/transaction.rs]: Unsafe code removed.

File: crates/engine/src/api/operation.rs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 use crate::EngineError;
 
 /// Storage operations
+/// TODO: refactor this trait, require `&mut self` for write operations
```
Comment (Phoenix500526) [inline crates/engine/src/api/operation.rs:4]: Please leave an issue to track.

Comment (bsbds) [reply crates/engine/src/api/operation.rs]: I think this TODO could be removed since rocksdb does not requires `&mut self`.

**Review by Phoenix500526 — APPROVED** [review]
Please squash commits

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 8]
```diff
@@ -1,18 +1,230 @@
...
+                    .txn
+                    .lock()
+                    .as_ref()
+                    .unwrap()
+                    .iterator_cf(&cf, mode)
+                    .take_while(|res| {
+                        res.as_ref()
+                            .is_ok_and(|(key, _)| key.as_ref() < to.as_slice())
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
```
Comment (rogercloud) [inline crates/engine/src/rocksdb_engine/transaction.rs:130]: If the kvs are iterated later, why are we collecting it?

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 9]
```diff
@@ -1,18 +1,230 @@
...
+            }
+        }
 
-impl TransactionApi for RocksTransaction {
-    fn commit(self) -> Result<(), crate::EngineError> {
         Ok(())
     }
 
-    fn rollback(&self) -> Result<(), crate::EngineError> {
+    fn write_multi(&self, ops: Vec<WriteOperation<'_>>, sync: bool) -> Result<(), EngineError> {
```
Comment (rogercloud) [inline crates/engine/src/rocksdb_engine/transaction.rs:140]: If we only iterate the ops only, shall we use IntoIterator trait here?

File: crates/engine/src/rocksdb_engine/transaction.rs (hunk 8 above)
Comment (bsbds) [reply crates/engine/src/rocksdb_engine/transaction.rs]: fixed

File: crates/engine/src/rocksdb_engine/transaction.rs (hunk 9 above)
Comment (bsbds) [reply crates/engine/src/rocksdb_engine/transaction.rs]: done

File: crates/engine/src/rocksdb_engine/transaction.rs [hunk 10]
```diff
@@ -1,18 +1,231 @@
...
+/// Transaction type for `RocksDB`
+pub struct RocksTransaction<'db> {
+    /// The inner DB
+    db: Arc<OptimisticTransactionDB>,
+    /// A transaction of the DB
+    ///
+    /// We need a `Mutex` because `Transaction<'db, DB>` does not implement `Sync`
+    txn: Mutex<Option<Transaction<'db, OptimisticTransactionDB>>>,
+    /// The size of the engine
+    engine_size: Arc<AtomicU64>,
```
Comment (Phoenix500526) [inline crates/engine/src/rocksdb_engine/transaction.rs:28]: The `AtomicU64` has already implemented `Sync` and `Send`. Why should we wrap it in a `Arc`? Can we use `&'db AtomicU64` to replace it?

**Comment by Phoenix500526** [comment]
Please squash the commits

File: crates/engine/src/rocksdb_engine/transaction.rs (hunk 10 above)
Comment (bsbds) [reply crates/engine/src/rocksdb_engine/transaction.rs]: done

**Review by Phoenix500526 — COMMENTED** [review]
LGTM

**Review by Phoenix500526 — APPROVED** [review]

**Pull request merged**